middle_end/flambda/simplify/env/rec_uses.ml

```diff
@@ -0,0 +1,337 @@
+          Name_occurrences.fold_variables name_occurrences ~init:graph
+            ~f:(fun g dst -> Var_graph.add_edge ~src ~dst g)
+        ) bindings graph
+      in
+      (* Build the graph of dependencies between continuation
+         parameters and arguments. *)
+      Continuation.Map.fold (fun k args (graph, used) ->
+        if Continuation.equal return_continuation k ||
+           Continuation.equal exn_continuation k then begin
+          let used =
+            Numbers.Int.Map.fold (fun _ name_occurrences used ->
+              add_used name_occurrences used
+            ) args used
+          in
+          graph, used
+        end else begin
+          let params =
+            match Continuation.Map.find k map with
+            | elt -> Array.of_list elt.params
+            | exception Not_found ->
+              Misc.fatal_errorf "Continuation not found during rec analysis: %a@."
+                Continuation.print k
+          in
+          let graph =
+            Numbers.Int.Map.fold (fun i name_occurrence graph ->
+              (* Note on the direction of the edge:
+                 We later do a reachability analysis to compute the
+                 transitive cloture of the used variables.
+                 Therefore an edge from src to dst means: if src is used, then
+                 dst is also used.
+                 Aplied here, this means : if the param of a continuation is used,
+                 then any argument provided for that param is also used.
+                 The other way wouldn't make much sense. *)
+              let src = params.(i) in
+              Name_occurrences.fold_variables name_occurrence ~init:graph
+                ~f:(fun g dst -> Var_graph.add_edge ~src ~dst g)
+            ) args graph
+          in
+          graph, used
+        end
+      ) apply_cont_args (graph, used)
+    ) map (Var_graph.empty, Variable.Set.empty)
+  in
+  (* Here we add to the graph the edges corresponding to the extra params
+     and args. *)
+  let graph =
+    Continuation.Map.fold (
+      fun _ (extra_params_and_args : Continuation_extra_params_and_args.t) graph ->
+      Apply_cont_rewrite_id.Map.fold (fun _ extra_args graph ->
+        List.fold_left2 (fun graph extra_param extra_arg ->
+          let src = Kinded_parameter.var extra_param in
+          match (extra_arg : Continuation_extra_params_and_args.Extra_arg.t) with
+          | Already_in_scope simple ->
+            Name_occurrences.fold_variables (Simple.free_names simple) ~init:graph
+              ~f:(fun g dst -> Var_graph.add_edge ~src ~dst g)
+          | New_let_binding (src', prim) ->
+            Name_occurrences.fold_variables (Flambda_primitive.free_names prim)
+              ~f:(fun g dst -> Var_graph.add_edge ~src:src' ~dst g)
+              ~init:(Var_graph.add_edge ~src ~dst:src' graph)
+        ) graph extra_params_and_args.extra_params extra_args
+      ) extra_params_and_args.extra_args graph
+    ) extra graph
+  in
```

Comment:
Maybe we could extract the graph building part into its own function to allow exposing the intermediate type to allow printing it for debug. This would add to the module interface
```OCaml
type dependency_graph = {
  dependency : Variable.Set.t Variable.Map.t;
  unconditionnaly_used : Variable.Set.t;
}
val build_dependency_graph : t -> dependency_graph
val print_graph : ppf -> dependency_graph -> unit
val required_variables : dependency_graph -> Variable.Set.t
```

Reply:
I split the code into a `Dependency_graph` submodule with your suggested type, the code is indeed slightly easier to read and understand now.

Reply:
Also: unconditionnaly_used -> unconditionally_used

Reply:
fixed.